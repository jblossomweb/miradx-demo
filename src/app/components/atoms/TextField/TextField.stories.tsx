import React from 'react'
import type { Meta, StoryObj } from '@storybook/react'

import TextField from './TextField'

type Story = StoryObj<typeof meta>

const meta = {
  title: 'components/atoms/TextField',
  component: TextField,
  render: (args) => <TextField {...args} />,
} satisfies Meta<typeof TextField>

export const Sample: Story = {
  args: {
    label: 'Sample TextField',
  },
}

export const Disabled: Story = {
  args: {
    label: 'Disabled TextField',
    disabled: true,
  },
}

export const Required: Story = {
  args: {
    label: 'Required TextField',
    required: true,
  },
}

export const Password: Story = {
  args: {
    label: 'Password TextField',
    type: 'password',
  },
}

export const Email: Story = {
  args: {
    label: 'Email TextField',
    type: 'email',
  },
}

export const Placeholder: Story = {
  args: {
    label: 'Sample TextField',
    placeholder: 'Sample Placeholder',
  },
}

export const NoLabel: Story = {
  args: {},
}

export const NoLabelPlaceholder: Story = {
  args: {
    placeholder: 'Sample Placeholder',
  },
}

export const NoLabelRequired: Story = {
  args: {
    required: true,
  },
}

export const FullWidth: Story = {
  args: {
    label: 'Full Width TextField',
    fullWidth: true,
  },
}

export const Controlled: Story = {
  args: {
    label: 'Controlled TextField',
    value: 'Controlled Value',
    onChange: (event) => {
      console.log(`set value to ${event.target.value}`)
    },
  },
}

export const ControlledPassword: Story = {
  args: {
    label: 'Password TextField',
    value: 'PasswordValue',
    type: 'password',
    onChange: (event) => {
      console.log(`set value to ${event.target.value}`)
    },
  },
}

export const WithError: Story = {
  args: {
    label: 'TextField With Error',
    error: 'Something is wrong with your input',
  },
}

export default meta
