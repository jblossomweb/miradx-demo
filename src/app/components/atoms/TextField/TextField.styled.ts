import tw from 'tailwind-styled-components'

export interface ContainerProps {
  $disabled?: boolean
}

export interface LabelProps {
  $required?: boolean
}

export interface InputProps {
  $fullWidth?: boolean
  $error?: boolean
}

export const Container = tw.label<ContainerProps>`
  flex
  flex-col
  items-left
  font-medium
  ${({ $disabled }) => ($disabled ? 'text-gray-400' : '')}
`

export const InputLabel = tw.span<LabelProps>`
  after:ml-1
  after:text-red-500
  ${({ $required }) => ($required ? `after:content-['*']` : '')}
`

export const Input = tw.input<InputProps>`
  ${({ $fullWidth }) => ($fullWidth ? 'w-full' : 'w-80')}
  h-8
  mr-2
  py-4
  px-1
  text-black
  rounded
  ring-1
  focus:ring-blue-500
  ${({ $error }) => ($error ? 'ring-red-500' : 'ring-transparent')}
  ${({ disabled }) => (disabled ? 'cursor-not-allowed' : '')}
  
`

export const InputError = tw.span`
  text-red-500
  font-bold
  italic
`
