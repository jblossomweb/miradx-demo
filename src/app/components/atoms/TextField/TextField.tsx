import React from 'react'
import * as Styled from './TextField.styled'

export interface TextFieldProps {
  label?: string
  disabled?: boolean
  placeholder?: string
  required?: boolean
  type?: 'email' | 'password' | 'text'
  fullWidth?: boolean
  value?: string
  error?: string
  onFocus?: (event: React.FocusEvent<HTMLInputElement>) => void
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
}

const TextField: React.FC<TextFieldProps> = ({
  label,
  disabled,
  placeholder,
  required,
  type = 'text',
  fullWidth,
  value,
  error,
  onFocus,
  onChange,
}) => (
  <Styled.Container $disabled={disabled}>
    {!!(label || required) && (
      <Styled.InputLabel $required={required}>{label}</Styled.InputLabel>
    )}
    <Styled.Input
      type={type}
      disabled={disabled}
      required={required}
      aria-label={label}
      value={value}
      placeholder={placeholder}
      onFocus={onFocus}
      onChange={onChange}
      $fullWidth={fullWidth}
      $error={!!error}
    />
    {!!error && <Styled.InputError>{error}</Styled.InputError>}
  </Styled.Container>
)

export default TextField
