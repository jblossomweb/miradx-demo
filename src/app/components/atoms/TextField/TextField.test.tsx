import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import storySnaps from 'test/utils/storySnaps'

import TextField from './TextField'
import * as stories from './TextField.stories'

describe('app/components/atoms/TextField', () => {
  describe('stories', () => {
    storySnaps(stories)
  })

  describe('assertions', () => {
    const onChange = jest.fn()
    const renderDefault = () => render(<TextField />)
    const renderRequired = () => render(<TextField required />)
    const renderLabel = (label: string) => render(<TextField label={label} />)
    const renderError = (error: string) => render(<TextField error={error} />)
    const renderPassword = (label: string) =>
      render(
        <TextField
          type="password"
          label={label}
        />,
      )
    const renderEmail = (label: string) =>
      render(
        <TextField
          type="email"
          label={label}
        />,
      )
    const renderControlled = (value: string) =>
      render(
        <TextField
          value={value}
          onChange={onChange}
        />,
      )

    beforeEach(jest.clearAllMocks)

    it('renders a text input', () => {
      renderDefault()
      const input = screen.getByRole('textbox')
      expect(input).toBeInTheDocument()
    })

    it('renders an empty text input by default', () => {
      renderDefault()
      const input = screen.getByRole('textbox') as HTMLInputElement
      expect(input.value).toBe('')
    })

    it('indicates to browser when text input is required', () => {
      renderRequired()
      const input = screen.getByRole('textbox') as HTMLInputElement
      expect(input.required).toBe(true)
    })

    it('indicates to browser when text input is a password', () => {
      const labelText = 'Password Field'
      renderPassword(labelText)
      const input = screen.getByLabelText(labelText) as HTMLInputElement
      expect(input.type).toBe('password')
    })

    it('indicates to browser when text input is an email', () => {
      const labelText = 'Email Field'
      renderEmail(labelText)
      const input = screen.getByLabelText(labelText) as HTMLInputElement
      expect(input.type).toBe('email')
    })

    it('renders a label when specified', () => {
      const labelText = 'Test Label'
      renderLabel(labelText)
      const label = screen.getByText(labelText)
      expect(label).toBeInTheDocument()
    })

    it('indicates label to screen readers when specified', () => {
      const labelText = 'Test Label'
      renderLabel(labelText)
      const input = screen.getByRole('textbox') as HTMLInputElement
      expect(input.getAttribute('aria-label')).toBe(labelText)
    })

    it('renders an error when specified', () => {
      const errorText = 'Test Error'
      renderError(errorText)
      const error = screen.getByText(errorText)
      expect(error).toBeInTheDocument()
    })

    it('controls input value when specified', () => {
      const valueText = 'Test Value'
      renderControlled(valueText)
      const input = screen.getByRole('textbox') as HTMLInputElement
      expect(input.value).toBe(valueText)
    })

    it('calls onChange when value is changed', () => {
      const initialValue = 'Initial Value'
      const newValue = 'New Value'
      const changeEvent = { target: { value: newValue } }
      renderControlled(initialValue)
      const input = screen.getByRole('textbox') as HTMLInputElement
      expect(input.value).toBe(initialValue)
      expect(onChange).not.toHaveBeenCalled()
      fireEvent.change(input, changeEvent)
      expect(onChange).toHaveBeenCalled()
    })
  })
})
