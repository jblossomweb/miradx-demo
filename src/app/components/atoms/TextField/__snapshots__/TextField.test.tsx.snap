// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/components/atoms/TextField stories Controlled matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Controlled TextField
    </span>
    <input
      aria-label="Controlled TextField"
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent"
      type="text"
      value="Controlled Value"
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories ControlledPassword matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Password TextField
    </span>
    <input
      aria-label="Password TextField"
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent"
      type="password"
      value="PasswordValue"
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories Disabled matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium text-gray-400"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Disabled TextField
    </span>
    <input
      aria-label="Disabled TextField"
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent cursor-not-allowed"
      disabled=""
      type="text"
      value=""
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories Email matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Email TextField
    </span>
    <input
      aria-label="Email TextField"
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent"
      type="email"
      value=""
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories FullWidth matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Full Width TextField
    </span>
    <input
      aria-label="Full Width TextField"
      class="w-full h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent"
      type="text"
      value=""
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories NoLabel matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <input
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent"
      type="text"
      value=""
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories NoLabelPlaceholder matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <input
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent"
      placeholder="Sample Placeholder"
      type="text"
      value=""
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories NoLabelRequired matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500 after:content-['*']"
    />
    <input
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent"
      required=""
      type="text"
      value=""
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories Password matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Password TextField
    </span>
    <input
      aria-label="Password TextField"
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent"
      type="password"
      value=""
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories Placeholder matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Sample TextField
    </span>
    <input
      aria-label="Sample TextField"
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent"
      placeholder="Sample Placeholder"
      type="text"
      value=""
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories Required matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500 after:content-['*']"
    >
      Required TextField
    </span>
    <input
      aria-label="Required TextField"
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent"
      required=""
      type="text"
      value=""
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories Sample matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      Sample TextField
    </span>
    <input
      aria-label="Sample TextField"
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-transparent"
      type="text"
      value=""
    />
  </label>
</DocumentFragment>
`;

exports[`app/components/atoms/TextField stories WithError matches markup snapshot 1`] = `
<DocumentFragment>
  <label
    class="flex flex-col items-left font-medium"
  >
    <span
      class="after:ml-1 after:text-red-500"
    >
      TextField With Error
    </span>
    <input
      aria-label="TextField With Error"
      class="w-80 h-8 mr-2 py-4 px-1 text-black rounded ring-1 focus:ring-blue-500 ring-red-500"
      type="text"
      value=""
    />
    <span
      class="text-red-500 font-bold italic"
    >
      Something is wrong with your input
    </span>
  </label>
</DocumentFragment>
`;
